core/embed/lib/error_handling.h

```diff
@@ -20,6 +20,151 @@
 #ifndef LIB_ERROR_HANDLING_H
 #define LIB_ERROR_HANDLING_H
 
+// Suppresses the intellisense error in VSCode
+#ifndef __FILE_NAME__
+#define __FILE_NAME__ __FILE__
+#endif
+
+// Build status code from any 16-bit value.
+// This macro is used for generic status codes.
+#define TS_BUILD(code) ((code) | (~(code) << 16))
+
+// Status codes
+//
+// Status code is hardened against fault injections
+// by storing the negated value in the upper 16 bits.
+typedef enum {
+  TS_OK = TS_BUILD(0),
+  TS_ERROR = TS_BUILD(1),          // Generic error
+  TS_ERROR_BUSY = TS_BUILD(2),     // Busy
+  TS_ERROR_TIMEOUT = TS_BUILD(3),  // Timeout
+  TS_ERROR_NOTINIT = TS_BUILD(4),  // Not initialized
+  TS_ERROR_ARG = TS_BUILD(4),      // Invalid argument
+  TS_ERROR_IO = TS_BUILD(5),       // I/O error
+} ts_t;
```

Comment:
I wonder what the set of error codes should be here. This particular set seems very specific to trezorhal. On the other hand trezor-crypto, for example, would use a completely different set of codes, like `TS_ERROR_SIGNATURE_INVALID`. Would this enum just be a large pile of error codes from various areas?

My impression is that if it is supposed to be a universal error type, then the set should be either very trivial (`sectrue`, `secfalse`) or maybe it should be something abstract like the protobuf `FailureType`.

Reply:
This approach is similar to the errno codes used in the libc - all codes are defined in one place. If you need a new error code, define it here, but try to come up with the most generic name so that someone else can use it as well.

However, we can change it. The error type does not need to be an enum; we can use `#define` and allow modules to define their own error codes. For example, the storage or optiga driver can then have specific error codes defined elsewhere.

Not sure what's better. Both ways have their own pros and cons.

Reply:
Yes, I agree. That's a good point. This approach eliminates a lot of bugs.

Reply:
I've changed `ts_t` to a struct in this commit cf075b5370487e7a88db7ad3295abd6196c1da4d. This change helped me to find some bugs that were introduced during  the sdcard driver refactoring...

Reply:
By changing the enum to an integer, we have lost the debugger's ability to show error codes in a readable form. However, we can easily fix this by using [Natvis](https://code.visualstudio.com/docs/cpp/natvis).

Reply:
I've also thought about module-specific error codes. Look at the notes below. In the end, we only need to add two new codes that are not so specific and can be used elsewhere:

One common set of error codes also allows us to create a function that returns an error name string.

```c
typedef enum _optiga_pin_result {
  OPTIGA_PIN_SUCCESS = 0,       // The operation completed successfully.
  OPTIGA_PIN_INVALID,           // The PIN is invalid.
  OPTIGA_PIN_COUNTER_EXCEEDED,  // The PIN try counter limit was exceeded.
  OPTIGA_PIN_ERROR,             // Optiga processing or communication error.
} optiga_pin_result;

OPTIGA_PIN_SUCCESS => TS_OK
OPTIGA_PIN_INVALID => TS_ERROR_ARG
OPTIGA_PIN_COUNTER_EXCEEDED => new error code: TS_ERROR_TRY_LIMIT
OPTIGA_PIN_ERROR => TS_ERROR_IO


typedef enum _optiga_sign_result {
  OPTIGA_SIGN_SUCCESS = 0,   // The operation completed successfully.
  OPTIGA_SIGN_INACCESSIBLE,  // The signing key is inaccessible.
  OPTIGA_SIGN_ERROR,         // Invalid parameters or Optiga processing or
                             // communication error.
} optiga_sign_result;

OPTIGA_SIGN_SUCCESS => TS_OK
OPTIGA_SIGN_INCACESSIBLE => new error code: TS_ERROR_SIGNING_KEY
OPTIGA_SIGN_ERROR => TS_ERROR_IO

```

Reply:
I am not a fan of mapping the codes like `OPTIGA_PIN_INVALID` -> `TS_ERROR_ARG`. That just seems error-prone, e.g. if some sub-function returns `TS_ERROR_ARG`, then we have to watch out to translate that to `OPTIGA_PIN_ERROR` or `TS_ERROR_IO`, to distinguish it from an invalid PIN. It's a pretty important distinction. If the PIN is invalid, then the user should be offered to retry, but if there is a runtime error, then we probably want the user to restart their device, rather than panic and deplete their PIN attempts.

Reply:
Ok, I understand. Maybe this case requires a more specific error code. We should also consider whether error codes should be used for situations like 'invalid PIN.' Is it really an error, or is it just invalid user input? Should be treated the same way as i/o error or invalid argument?